zephyr/include/sof/lib/regions_mm.h

```diff
@@ -10,31 +10,99 @@
+/* Dependency on ipc/topology.h created due to memory capability definitions
+ * that are defined there
+ */
+
+/* Defines maximum amount of memory block allocators in one heap
+ * since minimum realistic block should be DCASHE_LINE size
+ * and block sizes in allocator must be powers of 2
+ * so logicly it gives limited block sizes
+ * eg allocator of block size 64 128... 1024 8192
+ * and it is expected that allocations bigger than that would
+ * either be spaned on specific heap
  */
+#define MAXIMUM_BLOCK_SIZES 8
+
+/* list of virtual_memory_heap objects created */
+struct list_item vm_heaps;
+struct list_item physical_mappings;
+
 struct virtual_memory_heap {
+	struct list_item node;
 	/* zephyr provided virtual region */
 	struct sys_mm_drv_region *virtual_region;
-	/* physical pages allocators represented in memblocks */
-	struct sys_multi_mem_blocks physical_blocks_allocators;
-	/* SOF memory capability */
-	uint32_t memory_caps;
+	/* physical pages allocators represented in memblocks which
+	 * sizes are set on heap creation
+	 */
+	struct sys_mem_blocks *physical_blocks_allocators[MAXIMUM_BLOCK_SIZES];
+	/* keep current active allocations
+	 * TODO: further down the line it can be transitioned to memblocks
+	 * bitarray
+	 */
+	struct list_item active_allocations;
+	/* configuration flag marking if allocations will be contiguous */
+	bool span;
+};
+
+struct active_physical_mapping {
+	struct list_item node;
+
+	void *mapping_ptr;
+	uint32_t mapping_size;
 };
 
-/* Available externaly array containing all information on virtual heaps
- * Used to control physical allocations and overall virtual to physicall
- * mapping on sof side (zephyr handles the actual assigning physical memory
- * sof only requests it).
+struct active_allocation {
+	struct list_item node;
+
+	void *allocation_ptr;
+	uint32_t allocation_size;
+};
+
+struct block_bundle_descriptor {
+	uint32_t block_size;
+	uint32_t number_of_blocks;
+};
+/*
+ * Maybe this heap config should have small bundles first going up to max
+ * size or there should be a sorting mechanism for those ?
+ * Should we assume that bundles are given from smallest to bigest ?
+ */
+struct heap_config {
+	struct block_bundle_descriptor block_bundles_table[MAXIMUM_BLOCK_SIZES];
+};
+
+struct virtual_memory_heap *vmh_init_heap(const struct heap_config *cfg,
+	int memory_region_attribute, bool span);
+void *vmh_alloc(struct virtual_memory_heap *heap, size_t alloc_size);
+int vmh_free_heap(struct virtual_memory_heap *heap);
+int vmh_free(struct virtual_memory_heap *heap, void *ptr);
+struct virtual_memory_heap *vmh_reconfigure_heap(struct virtual_memory_heap *heap,
+	struct heap_config *cfg, bool span);
+struct heap_config vmh_get_default_heap_config(const struct sys_mm_drv_region *region);
+
+static inline bool is_ptr_in_range(void *ptr_1,
+	void *ptr_2, size_t size_2)
+{
+	return (uint8_t *)ptr_1 >= (uint8_t *)ptr_2 &&
+		   (uint8_t *)ptr_1 <= (uint8_t *)ptr_2 + size_2;
+}
```

Comment:
we should use ```vmh_``` prefix and use the needle, haystack, haystack_size naming as used by POSX for similar APIs.

Reply:
I changed the name of function but do  You mean internals as well ? Also I do not know the reference to needles and haystacks. I like to make names of functions understandable to everyone and above seems understandable to me.

Reply:
I just mean the external, - or even name ```is_ptr_in_range(void *range_start, size_t range_size, void *ptr)``` as we have some confusion when naming parameters with numbers